src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,3944 @@
+            (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
+
+        Returns:
+
+        """
+
+        logger.warning(
+            "This is the same forward method as `SeamlessM4TForTextToText`. It doesn't use `self.t2u_model`. If you want to generate speech, use the `generate` method."
+        )
+
+        return super().forward(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            decoder_input_ids=decoder_input_ids,
+            decoder_attention_mask=decoder_attention_mask,
+            head_mask=head_mask,
+            decoder_head_mask=decoder_head_mask,
+            cross_attn_head_mask=cross_attn_head_mask,
+            encoder_outputs=encoder_outputs,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            decoder_inputs_embeds=decoder_inputs_embeds,
+            labels=labels,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+    @torch.no_grad()
+    def generate(
+        self,
+        input_ids: Optional[torch.Tensor] = None,
+        return_intermediate_token_ids: Optional[bool] = None,
+        **kwargs,
+    ) -> Union[torch.Tensor, SeamlessM4TGenerationOutput]:
+        vocoder_tgt_lang_id = kwargs.pop("vocoder_tgt_lang_id", None)
+
+        kwargs_text = {"decoder_input_ids": kwargs.pop("decoder_input_ids", None)}
+        kwargs_speech = {}
+        for key, value in kwargs.items():
+            if key.startswith("text_"):
+                key = key[len("text_") :]
+                kwargs_text[key] = value
+            elif key.startswith("speech_"):
+                key = key[len("speech_") :]
+                kwargs_speech[key] = value
+            else:
+                # If the key is already in a specific config, then it's been set with a
+                # submodules specific value and we don't override
+                if key not in kwargs_text:
+                    kwargs_text[key] = value
+                if key not in kwargs_speech:
+                    kwargs_speech[key] = value
+
+        kwargs_text["output_hidden_states"] = True
+        kwargs_text["return_dict_in_generate"] = True
+        kwargs_text["output_scores"] = True
+
+        text_generation_output = super().generate(input_ids, **kwargs_text)
+
+        batch_size = len(input_ids)
+        num_return_sequences = len(text_generation_output.sequences) // batch_size
+        sequences = text_generation_output.sequences
+
+        attention_mask = kwargs_speech.get("attention_mask", kwargs_text.get("attention_mask", None))
+
+        encoder_hidden_states = text_generation_output.encoder_hidden_states[-1]
+
+        # compute last hidden state
+        t2u_input_embeds = self.text_decoder(
```

Comment:
After the text decoder the generation code gets a bit hacky - can we have a think about how we could make the high-level generation API intuitive for the user? For me, this means:
1. Unifying the kwarg prefixes across the different tasks (i.e. the encoder-decoder should always have the same prefix, the T2U model should always have the same prefix -> what prefixes can we give them so that they stay constant and intuitive across the models?)
2. Making the `language` / `speaker_id` arguments as intuitive as possible -> can we input these as `str` / `int` inputs respectively?

And for the lower level code:
1. Cleaning up the generation logic to make the token id shifts a bit clearer